smartsim/_core/mli/infrastructure/control/requestdispatcher.py

```diff
@@ -0,0 +1,515 @@
+        self._flush_lock = RLock()
+        """Lock used to make sure only one process can flush the queue (unused now)"""
+        self._uid = str(uuid.uuid4())
+        """Unique ID of queue"""
+
+    @property
+    def uid(self) -> str:
+        """ID of this queue"""
+        return self._uid
+
+    def acquire(self, blocking: bool = True, timeout: float = -1) -> t.Optional[bool]:
+        """Acquire queue lock to flush
+        :param blocking: whether to block on lock acquisition
+        :param timeout: Time to wait if blocking, before raising exception
+        """
+        return self._flush_lock.acquire(blocking=blocking, timeout=timeout)
+
+    def release(self) -> None:
+        """Release queue lock"""
+        self._flush_lock.release()
+
+    def __enter__(self) -> None:
+        """Method to use the Queue as a Context Manager"""
+        self.acquire()
+
+    def __exit__(
+        self,
+        exc_type: t.Optional[t.Type[BaseException]],
+        exc_val: t.Optional[BaseException],
+        exc_tb: t.Optional[TracebackType],
+    ) -> None:
+        """Method to release the Queue as a Context Manager"""
+        self.release()
+
+    @property
+    def model_key(self) -> str:
+        """Key of the model which needs to be run on the queued requests"""
+        return self._model_key
+
+    def put(
+        self,
+        item: InferenceRequest,
+        block: bool = False,
+        timeout: t.Optional[float] = 0.0,
+    ) -> None:
+        """Put an inference request in the queue
+        :param item: The request
+        :param block: Whether to block when trying to put the item
+        :param timeout: Time (in seconds) to wait if block==True
+        """
+        if not self.acquire(blocking=False):
+            raise Full
+        try:
+            if self.full():
+                raise Full
+            if self._first_put is None:
+                self._first_put = time.time()
+            super().put(item, block=block, timeout=timeout)
+        finally:
+            self.release()
+
+    @property
+    def _elapsed_time(self) -> float:
+        if self.empty():
+            return 0
+        return time.time() - self._first_put
+
+    @property
+    def ready(self) -> bool:
+        """True if the queue can be flushed"""
+        if self.empty():
+            return False
+        return self.full() or (self._elapsed_time >= self._batch_timeout)
+
+    def make_disposable(self) -> None:
+        """Set this queue as disposable, and never use it again after it gets flushed"""
+        self._disposable = True
+
+    @property
+    def can_be_removed(self) -> bool:
+        """Whether this queue can be deleted and garbafe collected"""
+        return self.empty() and self._disposable
+
+    def flush(self) -> list[t.Any]:
+        """Get all requests from queue
+        :return: Requests waiting to be executed
+        """
+        num_items = self.qsize()
```

Comment:
Should the lock on the queue be acquired when flushing?

Reply:
I removed the lock as we are not going to use this function in a multi-threaded environment. In the future, we might want to use a Dragon queue and a Dragon lock to go multi-processing